front/lib/swr/data_source_view_tables.ts

```diff
@@ -0,0 +1,199 @@
+    case "UPDATE":
+      return `${baseDataSourceUrl}/tables/${encodeURIComponent(tableId)}`;
+    case "READ":
+      return `${baseDataSourceViewUrl}/tables/${encodeURIComponent(tableId)}`;
+    case "READ_LIST":
+      return `${baseDataSourceViewUrl}/tables`;
+  }
+};
+
+export function useDataSourceViewTable({
+  dataSourceView,
+  tableId,
+  owner,
+  disabled,
+}: {
+  dataSourceView: DataSourceViewType | null;
+  tableId: string | null;
+  owner: LightWorkspaceType;
+  disabled?: boolean;
+}) {
+  const dataSourceViewTableFetcher: Fetcher<GetDataSourceViewTableResponseBody> =
+    fetcher;
+  const url =
+    dataSourceView && tableId
+      ? getUrl({
+          useCase: "READ",
+          owner,
+          dataSourceView,
+          tableId,
+        })
+      : null;
+
+  const { data, error, mutate } = useSWRWithDefaults(
+    url,
+    dataSourceViewTableFetcher,
+    {
+      disabled,
+    }
+  );
+
+  return {
+    table: data?.table,
+    isTableLoading: !disabled && !error && !data,
+    isTableError: error,
+    mutateTable: mutate,
+  };
+}
+
+export function useDataSourceViewTables({
+  dataSourceView,
+  owner,
+}: {
+  dataSourceView: DataSourceViewType | null;
+  owner: LightWorkspaceType;
+}) {
+  const tablesFetcher: Fetcher<ListTablesResponseBody> = fetcher;
+  const disabled = !dataSourceView;
+
+  const url = dataSourceView
+    ? getUrl({
+        useCase: "READ_LIST",
+        owner,
+        dataSourceView,
+      })
+    : null;
+  const { data, error, mutate } = useSWRWithDefaults(
+    disabled ? null : url,
+    tablesFetcher
+  );
+
+  return {
+    tables: useMemo(() => (data ? data.tables : []), [data]),
+    isTablesLoading: !disabled && !error && !data,
+    isTablesError: error,
+    mutateTables: mutate,
+  };
+}
+export function useUpdateDataSourceViewTable(
+  owner: LightWorkspaceType,
+  dataSourceView: DataSourceViewType,
+  tableName: string,
+  options?: SWRMutationConfiguration<PatchTableResponseBody, Error, string>
+) {
+  // Used only for cache invalidation
+  const { mutate: mutateContentNodes } = useDataSourceViewContentNodes({
+    owner,
+    dataSourceView,
+    disabled: true,
+  });
+
+  // Used only for cache invalidation
+  const { mutateTable } = useDataSourceViewTable({
+    owner,
+    dataSourceView,
+    tableId: tableName,
+    disabled: true,
+  });
+
+  // Decorate options's onSuccess with cache invalidation
+  const invalidateCacheEntries = async () => {
+    await Promise.all([mutateContentNodes, mutateTable]);
+  };
+  const decoratedOptions = decorateWithInvalidation(
+    options,
+    invalidateCacheEntries
+  );
+
+  const patchUrl = tableName
+    ? getUrl({
+        useCase: "UPDATE",
+        owner,
+        dataSourceView,
+        tableId: tableName,
+      })
+    : null;
+  const sendPatchRequest = mutationFn<PatchDataSourceTableRequestBody>("PATCH");
+  return useSWRMutation(patchUrl, sendPatchRequest, decoratedOptions);
```

Comment:
I took some time to re-read this to come up with a more factual comment.

Beyond the fact that it's a "new" pattern (vs existing), here is why I am not convinced by this version:
- It bypass the way we handle success or failures in general (via the `Result` system) in favor of callbacks
- OnSuccess / OnError handling (notification, returning a typed error) is delegated to the caller (fixable of course but at the expense of more meta code).
- It introduce more indirections, imho the code is less straightforward if we want to understand what happens (callbacks and getUrl() with a new "use case" concept above the rest API concepts, ...)
- Less flexible to use (pre-setup of the hook then mutation.trigger() vs doUpdate(name, description)).
- What if I don't have the tableName already ?
- What if I want args to my mutation function that are not directly in the payload ?

Compared to `useUpdateFolder` ([here](https://github.com/dust-tt/dust/blob/65a207b83ba7fda94bb07eb66b0d37fb9a020219/front/lib/swr/spaces.ts#L276)) who does the same thing (and more with the notifications etc..), I honestly find the code shorter, simpler, easier to use and more flexible if I need to evolve it.

Can you tell me why you are pushing to introduce this ? What benefits over the existing thing do you see that I might be overlooking ?

Reply:
For context some discussions around this pattern happened / are happening here: https://github.com/dust-tt/dust/pull/8751#discussion_r1856891201

Reply:
Hey @Fraggle, to answer your points:
- correct
- Not necessarily, that's wha I do with the decorateWithInvalidation  : run the onSuccess hook passed by the caller if any, but also mutate the cache -> swr-specific hook logic lives in swr 
- getUrl is somehow unrelated from useSWRMutation vs useSWR, it's a way to group keys to avoid scattering them across files -> makes it more maintainable imo. the useCase concept is an implication: if you impleent getUrl you need a handle on the url you want, can't be the method  because e.g. getting a table or a list of tables would both use GET but hit a different endpoint.
- Not necessarily true: afaik in both cases you've got two calls, one to get the trigger (or doUpdate) function, and one to actually trigger (doUpdate)
- Any parameter that's not available immediately during setup can be passed to trigger, either as a arg or in a options hook
- Not sure what side effect you'd want to trigger, but the args that you pass to trigger could be an object made of e.g. 'payload' and 'other_params', with payload only being passed to the fetch call

The reasons why I was pushing for it:
- useSWR also handles state management under the hood (isLoading,error, ...)
- You can pass hooks for onSuccess, onError
- Assuming you're posting/patching on the same url that should be invalidated (get), it's mutated under the hood

Now, I got general pushback and don't think it's worth changing all mutations to switch to useSWRMutation, so I am fine with reverting to a doUpdate-like pattern.

On the other hand, I think we'll eventually want to group keys in these getUrl methods for clarity, what do you think ?

In any case, thanks for the review ! :)

Reply:
Thanks for answering, I'll not comment on each individual points, especially if we are reverting to the generic pattern. My global point was to show that introducing a new pattern has a lot of ramifications that shouldn't be ignored / taken lightly.

> On the other hand, I think we'll eventually want to group keys in these getUrl methods for clarity, what do you think ?

I'm not sure of the value of the indirection created & the need to expose the implementation detail (aka: url = key).

I feel that as dev, I'd rather think of fetching hooks I need to invalidate/mutate than the underlying implementation.

For example: I create a hook to create a new document, so I need to mutate the hooks that list documents in general, the one that list document in the same datasource etc..). They might not even share the same url / key.